### ISSUE TITLE
Unoptimized query for locationlog latest entry

### ISSUE THREAD

**Issue by julsemaan** [inverse-inc/packetfence#3651]
The query below takes a lot of time when dealing with large node tables:
```
SELECT `locationlog`.`id`, `locationlog`.`tenant_id`, `locationlog`.`mac`, `locationlog`.`switch`, `locationlog`.`port`, `locationlog`.`vlan`, `locationlog`.`role`, `locationlog`.`connection_type`, `locationlog`.`connection_sub_type`, `locationlog`.`dot1x_username`, `locationlog`.`ssid`, `locationlog`.`start_time`, `locationlog`.`end_time`, `locationlog`.`switch_ip`, `locationlog`.`switch_mac`, `locationlog`.`stripped_user_name`, `locationlog`.`realm`, `locationlog`.`session_id`, `locationlog`.`ifDesc` FROM locationlog LEFT OUTER JOIN node ON ( `locationlog`.`mac` = `node`.`mac` ) WHERE ( ( `end_time` = '0000-00-00 00:00:00' AND `locationlog`.`tenant_id` = '1' AND `node`.`voip` = 'no' AND `port` = '40' AND `switch` = '192.168.5.3' ) ) ORDER BY `start_time` DESC
```

**Comment by julsemaan** [comment]
With @satkunas, we adjusted the query to:
```
SELECT 
  `locationlog`.`id`, 
  `locationlog`.`tenant_id`, 
  `locationlog`.`mac`, 
  `locationlog`.`switch`, 
  `locationlog`.`port`, 
  `locationlog`.`vlan`, 
  `locationlog`.`role`,
	`locationlog`.`connection_type`,
	`locationlog`.`connection_sub_type`,
	`locationlog`.`dot1x_username`,
	`locationlog`.`ssid`,
	`locationlog`.`start_time`,
	`locationlog`.`end_time`,
	`locationlog`.`switch_ip`,
	`locationlog`.`switch_mac`,
	`locationlog`.`stripped_user_name`,
	`locationlog`.`realm`,
	`locationlog`.`session_id`,
	`locationlog`.`ifDesc` 
FROM locationlog 
RIGHT JOIN node ON ( 
  `node`.`tenant_id` = 1
  AND `locationlog`.`mac` = `node`.`mac`
  AND `node`.`voip` = 'no' 
 ) 
WHERE (
  `locationlog`.`tenant_id` = '1' 
    AND `end_time` = '0000-00-00 00:00:00' 
    AND `port` = '40' 
    AND `switch` = '192.168.5.3'  
) ORDER BY `start_time` DESC
```

And got better results (no more temp table).

**Comment by julsemaan** [comment]
@jrouzierinverse, can you look at the new query and provide us your input on it, should provide the same result.

If that is fine, then we'd need to adjust the code in pf::locationlog::locationlog_view_open_switchport

**Comment by jrouzierinverse** [comment]
It should be joining where the tenant_id is equal for both node and locationlog instead of '1'.

```sql
SELECT 
  `locationlog`.`id`, 
  `locationlog`.`tenant_id`, 
  `locationlog`.`mac`, 
  `locationlog`.`switch`, 
  `locationlog`.`port`, 
  `locationlog`.`vlan`, 
  `locationlog`.`role`,
	`locationlog`.`connection_type`,
	`locationlog`.`connection_sub_type`,
	`locationlog`.`dot1x_username`,
	`locationlog`.`ssid`,
	`locationlog`.`start_time`,
	`locationlog`.`end_time`,
	`locationlog`.`switch_ip`,
	`locationlog`.`switch_mac`,
	`locationlog`.`stripped_user_name`,
	`locationlog`.`realm`,
	`locationlog`.`session_id`,
	`locationlog`.`ifDesc` 
FROM locationlog 
RIGHT JOIN node ON ( 
  `node`.`tenant_id` = `locationlog`.`tenant_id`
  AND `locationlog`.`mac` = `node`.`mac`
  AND `node`.`voip` = 'no' 
 ) 
WHERE (
    `end_time` = '0000-00-00 00:00:00' 
    AND `port` = '40' 
    AND `switch` = '192.168.5.3'  
) ORDER BY `start_time` DESC
```

**Comment by satkunas** [comment]
> It should be joining where the tenant_id is equal for both node and locationlog instead of '1'.

@jrouzierinverse agreed. The query still needs the `tenant_id` in the WHERE clause:

```sql
SELECT 
  `locationlog`.`id`, 
  `locationlog`.`tenant_id`, 
  `locationlog`.`mac`, 
  `locationlog`.`switch`, 
  `locationlog`.`port`, 
  `locationlog`.`vlan`, 
  `locationlog`.`role`,
	`locationlog`.`connection_type`,
	`locationlog`.`connection_sub_type`,
	`locationlog`.`dot1x_username`,
	`locationlog`.`ssid`,
	`locationlog`.`start_time`,
	`locationlog`.`end_time`,
	`locationlog`.`switch_ip`,
	`locationlog`.`switch_mac`,
	`locationlog`.`stripped_user_name`,
	`locationlog`.`realm`,
	`locationlog`.`session_id`,
	`locationlog`.`ifDesc` 
FROM locationlog 
RIGHT JOIN node ON ( 
  `node`.`tenant_id` = `locationlog`.`tenant_id`
  AND `node`.`mac` = `locationlog`.`mac`
  AND `node`.`voip` = 'no' 
 ) 
WHERE (
    `locationlog`.`tenant_id`=1
    AND `end_time` = '0000-00-00 00:00:00' 
    AND `port` = '40' 
    AND `switch` = '192.168.5.3'  
) ORDER BY `start_time` DESC
```

I would also change this

```sql
  AND `locationlog`.`mac` = `node`.`mac`
```

to this

```sql
  AND `node`.`mac` = `locationlog`.`mac`
```

.. in order to keep the JOINed table on the left side of the ON statement(s) (minor symantics).

**Comment by jrouzierinverse** [comment]
Yes the tenant_id =1  in the where clause needs to be there

**Comment by julsemaan** [comment]
I think that hasn't been said enough, but the tenant_id needs to be there :)

@jrouzierinverse, can you proceed with the implementation of the change in the locationlog module ?
That will need to go in the maintenance as well.

**Comment by julsemaan** [comment]
@jrouzierinverse 

Could you look into this so this gets into 8.2

**Comment by extrafu** [comment]
@julsemaan Please check if that was already fixed and the bug report isn't updated.

**Comment by julsemaan** [comment]
Already fixed in 2d8d0fa19da445bcc9071ce9dc8f0b0e756ceb4f and ea3aca98ab9b2c53cf8a9b1a798eed5d2f72c651
